src/datasets/features/features.py

```diff
@@ -1324,8 +1342,12 @@ def numpy_to_pyarrow_listarray(arr: np.ndarray, type: pa.DataType = None) -> pa.
     for i in range(arr.ndim - 1):
         n_offsets = reduce(mul, arr.shape[: arr.ndim - i - 1], 1)
         step_offsets = arr.shape[arr.ndim - i - 1]
-        offsets = pa.array(np.arange(n_offsets + 1) * step_offsets, type=pa.int32())
-        values = pa.ListArray.from_arrays(offsets, values)
+        if n_offsets * step_offsets < (1 << 31):
+            offsets = pa.array(np.arange(n_offsets + 1) * step_offsets, type=pa.int32())
+            values = pa.ListArray.from_arrays(offsets, values)
+        else:
+            offsets = pa.array(np.arange(n_offsets + 1) * step_offsets, type=pa.int64())
+            values = pa.LargeListArray.from_arrays(offsets, values)
```

Comment:
Have you tried using pa.chunked_array instead of pa.LargeListArray ? (i.e. chunking the input numpy array into small arrays to have a list of pa.ListArray that you concatenate with pa.chunked_array)

In the rest of the code base we don't support LargeListArray and so it could lead to issues when doing type inference or type casting

Reply:
Do you mean that it will be ok to return a pa.ChunkedArray?

Reply:
I think so but I haven't tested

Reply:
Oh yea indeed. I'm not sure it would even work if the shape was (1, 50000, 50000) since I believe a ListArray can't have a ChunkedArray as values.

Let's go for LargeListArray then, though we'll have to update a few things in table.py but this can be handled later I think (let me know if that's something you'd like to help with !)

Reply:
Sure! Could you provide some guidance?

Reply:
@lhoestq Hi, I checked `table.py` and found that `pa.ListArray` is only called in the `array_cast` function. Does that mean I only need to update this function for the adaption of LargeListArray?

Reply:
Yes indeed, in array_cast, cast_array_to_feature and embed_array_storage.

Then in features.py there's also generate_from_arrow_type to update

Finally there's still one situation that needs to be improved: when you concatenate a dataset with a list and a dataset with a large list. I think in this case the regular list should be promoted to a large list but this is maybe less trivial

Reply:
Do you mean the case where the two datasets to be concatenated share the same column? However, I think it would raise a ValueError for duplicated columns.

Reply:
And I've updated all these three functions, i.e., array_cast, cast_array_to_feature, and embed_array_storage.
Please review these changes as I am not familiar with these functions :)

Reply:
> Do you mean the case where the two datasets to be concatenated share the same column? However, I think it would raise a ValueError for duplicated columns.

When you have two datasets with the same column "embedding" but one has 50,000 elements (so it uses LargeListArray) and the second dataset has 1 element (so it uses a ListArray)

Reply:
take this as an example
```python
a = np.zeros((3,3))
da = datasets.Dataset.from_dict({"col": a})
b = np.ones((3,3))
db = datasets.Dataset.from_dict({"col": b})
res = datasets.concatenate_datasets([da, db], axis=1)
```

it would raise the following error
```none
Traceback (most recent call last):
  File "tmp.py", line 35, in <module>
    res = datasets.concatenate_datasets([da, db], axis=1)
  File "/data/wenjiaxin/home/datasets/src/datasets/combine.py", line 176, in concatenate_datasets
    return _concatenate_map_style_datasets(dsets, info=info, split=split, axis=axis)
  File "/data/wenjiaxin/home/datasets/src/datasets/arrow_dataset.py", line 4773, in _concatenate_map_style_datasets
    _check_column_names([col_name for dset in dsets for col_name in dset._data.column_names])
  File "/data/wenjiaxin/home/datasets/src/datasets/arrow_dataset.py", line 606, in _check_column_names
    raise ValueError(f"The table can't have duplicated columns but columns {duplicated_columns} are duplicated.")
ValueError: The table can't have duplicated columns but columns ['col'] are duplicated.
```

Reply:
I mean concatenate on axis=0 ;)

Reply:
@mariosasko @lhoestq Hi, do we have conclusions about this issue :)

Reply:
Maybe there's another way, what about using FixedSizeListArray instead ? It doesn't have an offsets array so it should fix the issue ?

Reply:
```python
a = np.zeros((50000, 50000))
da = datasets.Dataset.from_dict({"col": a}, features=datasets.Features({"col": Sequence(feature=Value(dtype='float64', id=None))}))
```

I test this case, and it successfully passes the sanity check. 

Both `self.info.features.type` and `inferred_features.type` is `struct<col: list<item: double>>`

Reply:
Oh yea right ! It returns an arrow type using `list` every time so it should be alright ! Good catch :)

Reply:
😄okay

Reply:
Considering that @mariosasko may be too busy these days, shall we merge this PR and support this important feature? @lhoestq

Reply:
we're discussing this internally and doing some tests to make sure - will keep you posted ;)

Reply:
From internal discussions, this code would fail because the ArrowWriter can't start writing large lists to disk if the schema has been determined to contain regular lists:
```python
import numpy as np
import datasets
a = np.zeros((5000000, 768), np.uint8)

datasets.Dataset.from_dict({"id": range(2)}).map(lambda _: {"a": a}, writer_batch_size=1, new_fingerprint="foo")
# ArrowInvalid: Array of type large_list<item: uint8> too large to convert to list<item: uint8>
```

This means that whenever someone runs `map` with varying lengths arrays and at one point an array doesn't fit into a regular ListArray, this will fail.

In this case we expect to let the user specify in advance that the largelist type must be used.
Therefore I think we need to add a parameter to `Sequence` to differentiate between regular lists and large lists. Maybe `Sequence(..., large=True)` ?
and this way we can also always verify
```python
schema == Features.from_arrow_schema(schema).arrow_schema
```

I hope that makes sense ^^'

Reply:
got it! Could you please provide some instructions on how to support this, e.g., functions I have to update

Reply:
I think the key is to add the `large` parameter to `Sequence` and update the functions you modified in this PR to use pa.list_() if large is False, and pa.large_list otherwise